src/Ydb.Sdk/src/Services/Query/Tx.cs

```diff
@@ -0,0 +1,167 @@
+            },
+            TxModeStaleSettings => new TransactionSettings
+            {
+                StaleReadOnly = new StaleModeSettings()
+            },
+            TxModeSnapshotSettings => new TransactionSettings
+            {
+                SnapshotReadOnly = new SnapshotModeSettings()
+            },
+            _ => throw new InvalidCastException(nameof(txModeSettings))
+        };
+        return txSettings;
+    }
+
+    public async Task<QueryResponseWithResult<T>> Query<T>(
+        string queryString,
+        Dictionary<string, YdbValue>? parameters,
+        Func<ExecuteQueryStream, Task<T>> func,
+        ExecuteQuerySettings? executeQuerySettings = null)
+    {
+        if (QueryClient is null)
+        {
+            throw new NullReferenceException($"{nameof(QueryClient)} is null");
+        }
+
+        if (SessionId is null)
+        {
+            throw new NullReferenceException($"{nameof(SessionId)} is null");
+        }
+
+        var stream = QueryClient.ExecuteQuery(SessionId, queryString, this, parameters,
+            executeQuerySettings);
+        try
+        {
+            var response = await func(stream);
+            return response is QueryClient.None
+                ? new QueryResponseWithResult<T>(Status.Success)
+                : new QueryResponseWithResult<T>(Status.Success, response);
+        }
+        catch (StatusUnsuccessfulException e)
+        {
+            return new QueryResponseWithResult<T>(e.Status);
+        }
+    }
+
+    public async Task<QueryResponseWithResult<T>> Query<T>(string queryString, Func<ExecuteQueryStream, Task<T>> func,
+        ExecuteQuerySettings? executeQuerySettings = null)
+    {
+        return await Query(queryString, new Dictionary<string, YdbValue>(), func, executeQuerySettings);
+    }
+
+    public async Task<QueryResponse> Exec(string queryString,
+        Dictionary<string, YdbValue>? parameters = null
+        , ExecuteQuerySettings? executeQuerySettings = null)
+    {
+        var response = await Query<QueryClient.None>(
+            queryString,
+            parameters,
+            async session =>
+            {
+                await QueryClient.EmptyStreamReadFunc(session);
+                return QueryClient.None.Instance;
+            },
+            executeQuerySettings);
+        return response;
+    }
+
+    public async Task<QueryResponseWithResult<Value.ResultSet.Row>> ReadFirstRow(string queryString,
+        Dictionary<string, YdbValue>? parameters = null,
+        ExecuteQuerySettings? executeQuerySettings = null)
+    {
+        var response = await Query(queryString, parameters, async stream =>
+            {
+                var rows = await QueryClient.ReadAllRowsHelper(stream);
+                return rows[0];
+            },
+            executeQuerySettings);
+        return response;
+    }
+
+    public async Task<QueryResponseWithResult<IReadOnlyList<Value.ResultSet.Row>>> ReadAllRows(string queryString,
```

Comment:
Need to check about result has parts with only one ResultSetIndex, for prevent combile rows from different statements.